[Lin Fangyuan] iP

LMAOboxhack opened the pull request:
# Bruh 
A *chatbot* for **21st century** task management.
## Testimonials
> Best app ~~this year~~ ever.
> – Jeve Stobs
## Features
* Task management
  * Example:
  * [x] Ace this mod
  * [ ] Take over the world 💪 
* Add & remove different types of tasks
...and many more!
## Installation
1. Download from [here](https://github.com/LMAOboxhack/ip/releases)
2. Double-click
3. Profit 💸 
## Printing a line in Java
Create a class & add the following `main` method:
```java
public class Main {
    public static void main(String[] args) {
        System.out.println("Hello world!");
    }
}
```

Commit:
Add Gradle support

LMAOboxhack committed:
Level 1

LMAOboxhack committed:
Level 2

LMAOboxhack committed:
Level 3

LMAOboxhack committed:
Level 4

LMAOboxhack committed:
Level 5

LMAOboxhack committed:
Level 6

LMAOboxhack committed:
A-TextUiTesting

LMAOboxhack committed:
Changed error handling

LMAOboxhack committed:
Add TaskCenter, TaskOperation classes

LMAOboxhack committed:
Add task list persistence

LMAOboxhack committed:
Add InputChecker, LocalDateTimeOrString classes

LMAOboxhack committed:
Add date & time parsing, utility classes

LMAOboxhack committed:
Add list by date and time command

LMAOboxhack committed:
Refactor and rename classes

LMAOboxhack committed:
Merge branch 'branch-Level-7'

LMAOboxhack committed:
Merge branch 'branch-Level-8'

LMAOboxhack committed:
Reorganize classes into packages

LMAOboxhack committed:
Add unit tests, gradle

LMAOboxhack committed:
Package app as JAR file

LMAOboxhack committed:
Add JavaDoc, rename classes

LMAOboxhack committed:
Update code to match coding standard

LMAOboxhack committed:
Add search tasks functionality, rename packages

LMAOboxhack committed:
Merge branch 'branch-A-JavaDoc'

LMAOboxhack committed:
Merge branch 'branch-A-CodingStandard'

LMAOboxhack committed:
Merge branch 'branch-Level-9'

LMAOboxhack committed:
Merge remote-tracking branch 'origin/add-gradle-support'

LMAOboxhack committed:
Modify package name in build.gradle

LMAOboxhack committed:
Merge branch 'branch-A-Gradle'

LMAOboxhack committed:
Add checkstyle, update formatting to match checkstyle

LMAOboxhack committed:
Merge branch 'branch-A-CheckStyle'

felissaf commented on src/main/java/bruh/tasklist/TaskList.java:
```diff
@@ -0,0 +1,128 @@
+     * @return A description of the number of tasks in the list.
+     */
+    public String getTaskCountDesc() {
+        return tasks.isEmpty() ? "There are no " + (isFiltered ? "matching " : "") + "tasks."
```
I believe it would be better to separate this into a few lines to avoid confusion!:smile: For the conditional expressions, maybe you can put it in a separate variable such as 
String plurality = (tasks.size() == 1) ? "" : "s";

LMAOboxhack replied:
Thanks for the suggestion! Agreed, it would be better to split this into multiple lines for more readability. Will fix!

felissaf commented on src/main/java/bruh/parser/Parser.java:
```diff
@@ -0,0 +1,88 @@
+        String[] inputs = input.split(" ", 2);
+        String keyword = inputs[0];
+
+        switch (keyword) {
```
I like the way you divide the different cases:smile:

felissaf commented on src/main/java/bruh/task/LocalDateTimeOrString.java:
```diff
@@ -0,0 +1,52 @@
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        } else if (o == null) {
```
Maybe it would be better to combine line 44 and 46 (o == null || !(o instanceof LocalDateTimeOrString)) as the return statement is the same?:smile:

LMAOboxhack replied:
Thanks for spotting this! Will fix.

felissaf commented on src/main/java/bruh/task/Task.java:
```diff
@@ -0,0 +1,85 @@
+ */
+public abstract class Task implements Serializable {
+    private final String description;
+    private final char symbol;
```
I believe constant variable should be in all uppercase?

LMAOboxhack replied:
Thanks for pointing this out! For these variables, they are not exactly constants; I added the `final` keyword for immutability, to reduce complexity and hopefully catch bugs, so I'm not too sure if these should be in all uppercase.

felissaf reviewed (commented):
You have very good code quality! I'm impressed:smile:

LMAOboxhack committed:
Add GUI, remove CLI functionality

LMAOboxhack committed:
Merge branch 'branch-Level-10'

Xiaoyunnn commented on src/main/java/bruh/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,29 @@
+    private final Task taskToAdd;
+
+    /**
+     * Constructor for a command to add a task to the task list.
```
According to the coding standards, I suppose the first sentence should start with something like "Instantiates" or "Constructs" instead of "Constructor...".

LMAOboxhack replied:
Thanks for pointing this out! I took a look at the coding standards and this doesn't seem to be explicitly specified, so I went on to take a look at the [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html#s7.2-summary-fragment) (the coding standard recommends this for topics not covered) and it seems that it is okay for the summary fragment of the Javadoc to begin with a noun phrase.

Xiaoyunnn commented on src/main/java/bruh/parser/Parser.java:
```diff
@@ -0,0 +1,88 @@
+        String keyword = inputs[0];
+
+        switch (keyword) {
+        case "todo":
+        case "deadline":
```
According to the coding standards,

> The explicit //Fallthrough comment should be included whenever there is a case statement without a break statement.

So perhaps, you may want to add `// Fallthrough` at the end of line 43?

LMAOboxhack replied:
Thanks for catching this! Will fix.

Xiaoyunnn commented on src/main/java/bruh/parser/Parser.java:
```diff
@@ -0,0 +1,88 @@
+     * @return The command to be run by the chatbot.
+     */
+    public static Command parseInputToCommand(String input) throws BruhException {
+        if (input.isEmpty()) {
+            throw new InvalidArgumentException();
+        } else if (input.equals(STOP_SIGNAL)) {
+            return new ExitCommand();
+        } else if (input.equals("list")) {
+            return new ListTasksCommand();
+        }
```
Perhaps you can consider adding the `else` branch to detect errors as mentioned in the code quality section of the textbook [here](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#use-the-default-branch).

LMAOboxhack replied:
Thanks for raising this up! My rationale for not adding an else branch here is that I want to allow the control to simply proceed to the subsequent `switch` statements, where I included a `default` case. I believe this will catch any inputs that do not match any of the cases & simply throw an `InvalidArgumentException`.

Xiaoyunnn commented on src/main/java/bruh/parser/Parser.java:
```diff
@@ -0,0 +1,88 @@
+     *
+     * @param sections     An array consisting of the input, split into the first word and the rest of the input
+     * @param errorMessage The error message to be displayed, if an exception is thrown.
+     * @throws MissingArgumentException if a missing argument is found.
```
I think you missed out the upper case for "if a missing ..." according to the coding standard XD

LMAOboxhack replied:
Thanks for pointing this out! Will fix.

michael-lee-sk commented on src/main/java/bruh/parser/Parser.java:
```diff
@@ -0,0 +1,88 @@
+
+        switch (keyword) {
+        case "todo":
+        case "deadline":
```
Agree with @Xiaoyunnn on adding a //Fallthrough comment

LMAOboxhack replied:
Thanks for pointing this out! Will fix.

michael-lee-sk commented on src/main/java/bruh/task/LocalDateTimeOrString.java:
```diff
@@ -0,0 +1,52 @@
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        } else if (o == null) {
```
Agree with @felissaf, that you should just merge both else if statements, unless you intend to print a different statement for each case.

LMAOboxhack replied:
Thanks for raising this! Agreed, will merge both cases.

michael-lee-sk reviewed (commented):
Didn't have much to comment on on top of @Xiaoyunnn and @felissaf. Thought u have pretty clean code 👍

LMAOboxhack committed:
Delete dead code

JUnit tests as well as some scripts used for testing the previous
text-based UI still remain in the code.

These files are redundant and create clutter.

Deleting these files will reduce such clutter and
make our code cleaner.

LMAOboxhack committed:
Refactor code to match code quality standards

Code contains violations of code quality standards, such as
use of magic literals and complicated expressions.

This can reduce readability and maintainability.

Various improvements have been made to improve code quality,
such as extracting magic literals to constants, and extracting
code to separate methods.

This is done in accordance with the resources and materials
provided in the CS2103 module with regards to code quality.

LMAOboxhack committed:
Incorporate Java streams into task list filtering

LMAOboxhack committed:
Update gitignore

LMAOboxhack committed:
Add assertions for error checking

LMAOboxhack committed:
Merge branch 'master' into branch-A-CodeQuality

LMAOboxhack committed:
Merge branch 'master' into branch-A-Streams

LMAOboxhack committed:
Merge pull request #3 from LMAOboxhack/branch-A-CodeQuality

Refactor code to match code quality standards

LMAOboxhack committed:
Update gitignore

LMAOboxhack committed:
Merge branch 'master' into branch-A-Assertions

LMAOboxhack committed:
Merge pull request #1 from LMAOboxhack/branch-A-Assertions

Add assertions for error checking

LMAOboxhack committed:
Merge branch 'master' into branch-A-Streams

LMAOboxhack committed:
Untrack files in gitignore

LMAOboxhack committed:
Merge pull request #2 from LMAOboxhack/branch-A-Streams

Incorporate Java streams into task list filtering

LMAOboxhack committed:
Add file for GitHub Actions

LMAOboxhack committed:
Change gradlew permission

LMAOboxhack committed:
Replace CRLF in gradlew with LF

LMAOboxhack committed:
Add support for 'Do After' tasks, refactor tasks

LMAOboxhack committed:
Merge branch 'branch-B-DoAfterTasks'

LMAOboxhack committed:
Fix switch/case indentation

LMAOboxhack committed:
Set theme jekyll-theme-cayman

LMAOboxhack committed:
Add representative screenshot, update profile picture

LMAOboxhack committed:
Update README with user guide

LMAOboxhack committed:
Update README formatting

LMAOboxhack committed:
Give credit for reused work
